src/veh_interact.cpp

```diff
@@ -934,79 +927,94 @@ void veh_interact::do_mend()
 
 }
 
-/**
- * Handles refilling a vehicle's fuel tank.
- * @param reason INVALID_TARGET if there's no fuel tank in the spot,
- *               CANT_REFILL All tanks are broken or player has nothing to fill the tank with.
- */
 void veh_interact::do_refill()
 {
-    const task_reason reason = cant_do('f');
-    display_mode('f');
-    werase (w_msg);
-    int msg_width = getmaxx(w_msg);
+    display_mode( 'f' );
+    werase( w_msg );
 
-    switch (reason) {
-    case INVALID_TARGET:
-        mvwprintz(w_msg, 0, 1, c_ltred, _("There's no refillable part here."));
-        wrefresh (w_msg);
-        return;
-    case CANT_REFILL:
-        mvwprintz(w_msg, 1, 1, c_ltred, _("There is no refillable part here."));
-        mvwprintz(w_msg, 2, 1, c_white, _("The part is broken, or you don't have the "
-                                          "proper fuel."));
-        wrefresh (w_msg);
-        return;
-    case MOVING_VEHICLE:
-        fold_and_print( w_msg, 0, 1, msg_width - 2, c_ltgray,
-                        _( "You can't refill the vehicle while driving." ) );
-        wrefresh (w_msg);
-        return;
-    default:
-        break; // no reason, all is well
+    // get all tanks and determine if possible to refill with currently available items
+    std::vector<std::pair<vehicle_part *, bool>> tanks;
+    for( auto &pt : veh->parts ) {
+        if( pt.is_tank() && !pt.is_broken() ) {
+
+            auto func = [&]( const item &e ) {
+                // cannot refill using active liquids (those that rot) due to #18570
+                if( e.is_watertight_container() && !e.contents.empty() && !e.contents.front().active ) {
+                    return pt.can_reload( e.contents.front().typeId() );
+                }
+                return false;
+            };
+
+            tanks.emplace_back( &pt, g->u.has_item_with( func ) ||
+                                     map_selector( g->u.pos(), 1 ).has_item_with( func ) ||
+                                     vehicle_selector( g->u.pos(), 1 ).has_item_with( func ) );
```

Comment:
Not fully related, but a wrapper selector for "all available items nearby" would be nicer than this big block.